jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRobustnessSuite.scala

```diff
@@ -0,0 +1,136 @@
+import org.apache.spark.{SparkConf, SparkContext}
+import org.scalatest.FunSuite
+
+class XGBoostRobustnessSuite extends FunSuite with Utils {
+  test("test Java RabitTracker wrapper's exception handling: it should not hang forever.") {
+    // Explicitly create new instances of SparkContext in each test to avoid reusing the same
+    // thread pool, which corrupts the internal state of Rabit and causes crash.
+    val sparkConf = new SparkConf().setMaster("local[*]")
+      .setAppName("XGBoostSuite").set("spark.driver.memory", "512m")
+    implicit val sparkContext = new SparkContext(sparkConf)
+    sparkContext.setLogLevel("ERROR")
+
+    val rdd = sparkContext.parallelize(1 to numWorkers, numWorkers).cache()
+
+    val tracker = new PyRabitTracker(numWorkers)
+    tracker.start(0)
+    val trackerEnvs = tracker.getWorkerEnvs
+
+    val dummyTasks = rdd.mapPartitions { iter =>
+      Rabit.init(trackerEnvs)
+      Thread.sleep(500)
+      val index = iter.next()
+      if (index == 1) {
+        // kill the worker by throwing an exception
+        throw new RuntimeException("Worker exception.")
+      }
+      Rabit.shutdown()
+      Iterator(index)
+    }.cache()
+
+    val sparkThread = new Thread() {
+      override def run(): Unit = {
+        // forces a Spark job.
+        dummyTasks.foreachPartition(() => _)
+      }
+    }
+
+    sparkThread.setUncaughtExceptionHandler(tracker)
+    sparkThread.start()
+    assert(tracker.waitFor(0) != 0)
+    sparkContext.stop()
+  }
+
+  test("test Scala RabitTracker's exception handling: it should not hang forever.") {
+    val sparkConf = new SparkConf().setMaster("local[*]")
+      .setAppName("XGBoostSuite").set("spark.driver.memory", "512m")
+    implicit val sparkContext = new SparkContext(sparkConf)
+    sparkContext.setLogLevel("ERROR")
+
+    val rdd = sparkContext.parallelize(1 to numWorkers, numWorkers).cache()
+
+    val tracker = new ScalaRabitTracker(numWorkers)
+    tracker.start(0)
+    val trackerEnvs = tracker.getWorkerEnvs
+
+    val dummyTasks = rdd.mapPartitions { iter =>
+      Rabit.init(trackerEnvs)
+      Thread.sleep(500)
+      val index = iter.next()
+      if (index == 1) {
+        // kill the worker by throwing an exception
+        throw new RuntimeException("Worker exception.")
+      }
+      Rabit.shutdown()
+      Iterator(index)
+    }.cache()
+
+    val sparkThread = new Thread() {
+      override def run(): Unit = {
+        // forces a Spark job.
+        dummyTasks.foreachPartition(() => _)
+      }
+    }
+    sparkThread.setUncaughtExceptionHandler(tracker)
+    sparkThread.start()
+    assert(tracker.waitFor() == ScalaRabitTracker.FAILURE.statusCode)
+    sparkContext.stop()
+  }
+
+  test("test Scala RabitTracker's workerConnectionTimeout") {
+    val sparkConf = new SparkConf().setMaster("local[*]")
+      .setAppName("XGBoostSuite").set("spark.driver.memory", "512m")
+    implicit val sparkContext = new SparkContext(sparkConf)
```

Comment:
See the comment for creating the first `sparkContext`. I did initially extend `SharedSparkContext`, and realized that if I reuse the Spark context, it will reuse the same Thread pool locally. Because I am literally simulating worker exception here, it means that `Rabit.shutdown()` will never be called, hence the internal pointer is not released and set to `nullptr` in the C++ code.

As a side effect, you can't call `Rabit.init()` twice within the same thread. I explicitly create new Spark context to force creating a new Thread pool.